libs/vkd3d/command_queue_vkd3d_ext.c

```diff
@@ -0,0 +1,91 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#define VKD3D_DBG_CHANNEL VKD3D_DBG_CHANNEL_API
+
+#include "vkd3d_private.h"
+
+static inline struct d3d12_command_queue *d3d12_command_queue_from_ID3D12CommandQueueExt(d3d12_command_queue_vkd3d_ext_iface *iface)
+{
+    return CONTAINING_RECORD(iface, struct d3d12_command_queue, ID3D12CommandQueueExt_iface);
+}
+
+extern ULONG STDMETHODCALLTYPE d3d12_command_queue_AddRef(d3d12_command_queue_iface *iface);
+
+ULONG STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_AddRef(d3d12_command_queue_vkd3d_ext_iface *iface)
+{
+    struct d3d12_command_queue *command_queue = d3d12_command_queue_from_ID3D12CommandQueueExt(iface);
+    return d3d12_command_queue_AddRef(&command_queue->ID3D12CommandQueue_iface);
+}
+
+extern ULONG STDMETHODCALLTYPE d3d12_command_queue_Release(d3d12_command_queue_iface *iface);
+
+static ULONG STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_Release(d3d12_command_queue_vkd3d_ext_iface *iface)
+{
+    struct d3d12_command_queue *command_queue = d3d12_command_queue_from_ID3D12CommandQueueExt(iface);
+    return d3d12_command_queue_Release(&command_queue->ID3D12CommandQueue_iface);
+}
+
+extern HRESULT STDMETHODCALLTYPE d3d12_command_queue_QueryInterface(d3d12_command_queue_iface *iface,
+        REFIID iid, void **object);
+
+static HRESULT STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_QueryInterface(d3d12_command_queue_vkd3d_ext_iface *iface,
+        REFIID iid, void **out)
+{
+    struct d3d12_command_queue *command_queue = d3d12_command_queue_from_ID3D12CommandQueueExt(iface);
+    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
+    return d3d12_command_queue_QueryInterface(&command_queue->ID3D12CommandQueue_iface, iid, out);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_NotifyOutOfBandCommandQueue(d3d12_command_queue_vkd3d_ext_iface *iface, D3D12_OUT_OF_BAND_CQ_TYPE type)
+{
+    struct d3d12_command_queue* command_queue;
+    int i;
+
+    command_queue = d3d12_command_queue_from_ID3D12CommandQueueExt(iface);
+
+    if (!command_queue->device->vk_info.NV_low_latency2)
+        return E_NOTIMPL;
+
+    if (type != OUT_OF_BAND_RENDER && type != OUT_OF_BAND_PRESENT)
+        return E_INVALIDARG;
+
+    for (i = 0; i < VKD3D_QUEUE_FAMILY_COUNT; i++)
+    {
+        if (command_queue->device->queue_families[i]->vk_family_index == command_queue->vkd3d_queue->vk_family_index &&
+                command_queue->device->queue_families[i]->out_of_band_queue)
+        {
+            command_queue->vkd3d_queue = command_queue->device->queue_families[i]->out_of_band_queue;
```

Comment:
This breaks if NotifyOutOfBandCommandQueue is called after any submission to the queue has been made. Is this only intended to be called from the SDK right after queue creation?

Reply:
I would expect applications to mark the command queue as being out of band before submitting any work to it. 

Why would switching to another VkQueue in the same queue family cause issues here? I guess any barriers contained in work submitted after the switch won't properly wait for work submitted on the previous queue, but that could be fixed with a call to vkQueueWaitIdle. Is there a vkd3d-proton implementation detail I am missing as to why this wouldn't work?

Reply:
> I would expect applications to mark the command queue as being out of band before submitting any work to it.

Does the NvAPI extension specify or document that? I don't expect this to cause issues, but it needs to be clarified somewhere.

> Why would switching to another VkQueue in the same queue family cause issues here?

Because ExecuteCommandLists queues up work to the submit thread. It will access vkd3d_queue there. There's also the rule of implicit sync between queues in D3D12. ExecuteCommandLists on the same ID3D12Queue are implicitly synchronized. Swapping the queue mid-flight breaks that.

Reply:
Ok, going to assume that calling NotifyOOB after submitting work to it is NvAPI or app bug and move on.